[IMP] Convert XMLRPC handler to a regular HTTP controller

Yajo opened the pull request:
Description of the issue/feature this PR addresses: This way, XMLRPC calls (and, BTW, normal HTTP calls) can have a way to know i.e. the remote IP.

Current behavior before PR: See #24183

Desired behavior after PR is merged: Fix #24183

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr
@Tecnativa

Yajo commented:
I did this agains master, but it would be completely harmless to v9+. Would you accept it there? :thinking:

pedrobaeza commented:
@Yajo, there's no need for opening an issue and then put the PR. Put here all the description of the issue instead.

pedrobaeza commented:
@odony this comes due to our audit we are doing for GDPR compliance, so I think it's something that you have to check and cover in all supported versions (even in v9, as GDPR starts in May). We are completing this security risk with the module auth_brute_force with this change: https://github.com/OCA/server-tools/pull/1219.

There are others like password_security (https://github.com/OCA/server-tools/tree/10.0/password_security) that we have also as another security risk, although with OAuth server, the rules change, and the implementation of the robustness of the password should be done on that side.

amigrave commented:
Hi @Yajo,

Could you help me understand the reason why the needed information from the wsgi environment (REMOTE_ADDR in this case) can't be passed to the models from the controllers/xmlrpc layer entry points ?

Yajo commented:
It's a little bit tricky but I'll try :sweat_smile: 

1. An HTTP request reaches our WSGI server. It contains the `environ`, which holds all the information for the request: https://github.com/odoo/odoo/blob/7b49f5836c0f4a491631d27acb74ddd0106c86b1/odoo/service/wsgi_server.py#L166
1. The request is then handled by the XMLRPC system if possible, and if it's not possible then it's handled by the HTTP system; so there are 2 possibilities: https://github.com/odoo/odoo/blob/7b49f5836c0f4a491631d27acb74ddd0106c86b1/odoo/service/wsgi_server.py#L153-L157
   1. The first one is when **it's actually a valid XMLRPC** controller. In such case, the `wsgi_xmlrpc` method gets the `environ` parameter, but that one gets lost in this method's scope when it calls `odoo.http.dispatch_rpc`, which actually redirects the request to the corresponding service. From there downwards in the stack, there's no more way to get the environment details: https://github.com/odoo/odoo/blob/7b49f5836c0f4a491631d27acb74ddd0106c86b1/odoo/service/wsgi_server.py#L102-L124
   1. The 2nd one is when **it's not an XMLRPC  request**. Then it gets handled by the `odoo.http.root` app, which goes down the rabbit hole until the method that sets up the request: https://github.com/odoo/odoo/blob/7b49f5836c0f4a491631d27acb74ddd0106c86b1/odoo/http.py#L1448
   After this, we have the request set up in place, but it also it keeps on going down through `Model(ir.http)._dispatch()`, `Model(ir.http)._find_handler()`, `odoo.http.routing_map().bind_to_environ()`, and some others (also depending on installed addons), and finally the target controller.

And that's basically why XMLRPC requests cannot know anything from their WSGI environment.

The patch I'm proposing in this PR is quite simple: always attach the environ dict to the current thread, which is an object that you can always get. Indeed, not the most elegant solution, but would serve as a workaround.

In an ideal world, IMHO `wsgi_xmlrpc` should be dumped (possibly along with `odoo.http.dispatch_rpc`, `odoo.service.{common,db,model,security,server}`), and all requests should go the 2.ii way. XMLRPC requests should just be some extra controllers defined in `odoo/odoo/addons/base/controllers/main.py`, so when the requests get there, they are processed as normal controllers, having the ability to import `odoo.http.request` and work with it as normal.

Of course, `master` would be the place for that refactoring; that's the reasoning behind https://github.com/odoo/odoo/pull/24187#issuecomment-380445287.

odony commented:
Just a word about the legal rationale here. GDPR requirements apply to businesses, not to software. There are a great number of ways to implement security best practices and things like password security, logging, and auditing. None of them are imposed by the law. DoS and brute-force protection are good candidates for deployment at the firewall or web server level (nginx/apache) too, for example, just like HTTPS, instead of the application level. GDPR[1] should not be used as an excuse for ugly patches ;-)

Regarding this PR, it'd be much better to work on the conversion of XMLRPC to a normal controller (in master). For your immediate need you could use a monkey-patch. Or just disable XML-RPC entirely in favor of JSON-RPC (a normal controller). Again, this can be done at the application, web server, or firewall level too. If you can afford to do it, blocking XML-RPC sounds even safer than increasing its logging.

[1] "Gore Development Pull Request", according to @amigrave ;-)

Yajo commented:
Thanks @odony. According to your comments:

- I added monkey-patching to https://github.com/OCA/server-tools/pull/1219 so we can continue development there
- I refactored this PR to make XMLRPC handlers land in normal controllers.

Please notice that I haven't a proper `master` development environment set up, so I couldn't run tests. I hope all goes green and this code serves to you as a base to add this feature :sweat_smile: 

---


> DoS and brute-force protection are good candidates for deployment at the firewall or web server level (nginx/apache) too, for example, just like HTTPS, instead of the application level.

You're right. I just wanted to point out that Odoo currently cannot plug into these systems such as fail2ban or similars because:

1. Cannot know remote IP address (the problem you already know)
1. You could hit #18052 if you don't upgrade Werkzeug to 0.14.1, resulting in banning wrong IPs (and possibly blocking Odoo proxy, thus all users).
1. Odoo only logs auth attempt failures when the database/user combination exists. Brute force attacks that try to even guess the user name would pass unnoticed.

Yajo commented:
I can't believe I passed the tests! :eyes: :heavy_check_mark:

odony commented:
@Yajo thanks! :-)

@antonylesuisse I think you wanted to check this out

antonylesuisse commented:
Looks good but:
- i would move controller in http next to jsonrpc.
- we need to fix useless session creation.
